syntax = "proto3";
import "storage.proto";

package PB;
option java_package = "com.flipperdevices.protobuf";

enum CommandStatus {
    OK = 0;

    /**< Common Errors */
    ERROR = 1; /**< Unknown error */
    ERROR_DECODE = 2; /**< Command can't be decoded successfully - command_id in response may be wrong! */
    ERROR_NOT_IMPLEMENTED = 3; /**< Command succesfully decoded, but not implemented (deprecated or not yet implemented) */
    ERROR_BUSY = 4; /**< Somebody took global lock, so not all commands are available */
    ERROR_CONTINUOUS_COMMAND_INTERRUPTED = 14; /**< Not received has_next == 0 */
    ERROR_INVALID_PARAMETERS = 15; /**< not provided (or provided invalid) crucial parameters to perform rpc */

    /**< Storage Errors */
    ERROR_STORAGE_NOT_READY = 5; /**< FS not ready */
    ERROR_STORAGE_EXIST = 6; /**< File/Dir alrady exist */
    ERROR_STORAGE_NOT_EXIST = 7; /**< File/Dir does not exist */
    ERROR_STORAGE_INVALID_PARAMETER = 8; /**< Invalid API parameter */
    ERROR_STORAGE_DENIED = 9; /**< Access denied */
    ERROR_STORAGE_INVALID_NAME = 10; /**< Invalid name/path */
    ERROR_STORAGE_INTERNAL = 11; /**< Internal error */
    ERROR_STORAGE_NOT_IMPLEMENTED = 12; /**< Functon not implemented */
    ERROR_STORAGE_ALREADY_OPEN = 13; /**< File/Dir already opened */
    ERROR_STORAGE_DIR_NOT_EMPTY = 18; /**< Directory, you're going to remove is not empty */

    /**< Application Errors */
    ERROR_APP_CANT_START = 16; /**< Can't start app - internal error */
    ERROR_APP_SYSTEM_LOCKED = 17; /**< Another app is running */
    ERROR_APP_NOT_RUNNING = 21; /**< App is not running or doesn't support RPC commands */
    ERROR_APP_CMD_ERROR = 22; /**< Command execution error */

    /**< Virtual Display Errors */
    ERROR_VIRTUAL_DISPLAY_ALREADY_STARTED = 19; /**< Virtual Display session can't be started twice */
    ERROR_VIRTUAL_DISPLAY_NOT_STARTED = 20; /**< Virtual Display session can't be stopped when it's not started */

    /**< GPIO Errors */
    ERROR_GPIO_MODE_INCORRECT = 58;
    ERROR_GPIO_UNKNOWN_PIN_MODE = 59;
}

/* There are Server commands (e.g. Storage_write), which have no body message
 * in response. But 'oneof' obligate to have at least 1 encoded message
 * in scope. For this needs Empty message is implemented.
 */
message Empty {
}

message StopSession {
}

message Main {
    uint32 command_id = 1;
    CommandStatus command_status = 2;
    bool has_next = 3;
    oneof content {
        Empty empty = 4;
        StopSession stop_session = 19;
        .PB_Storage.InfoRequest storage_info_request = 28;
        .PB_Storage.InfoResponse storage_info_response = 29;
        .PB_Storage.ListRequest storage_list_request = 7;
        .PB_Storage.ListResponse storage_list_response = 8;
        .PB_Storage.ReadRequest storage_read_request = 9;
        .PB_Storage.ReadResponse storage_read_response = 10;
        .PB_Storage.WriteRequest storage_write_request = 11;
        .PB_Storage.DeleteRequest storage_delete_request = 12;
    }
}

message Region {
    message Band {
        uint32 start = 1;
        uint32 end = 2;
        int32 power_limit = 3;
        uint32 duty_cycle = 4;
    }

    bytes country_code = 1;
    repeated Band bands = 2;
}